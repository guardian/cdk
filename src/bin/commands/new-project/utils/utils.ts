import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
