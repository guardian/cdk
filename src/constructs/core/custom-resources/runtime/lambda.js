/*
* The reason for this file is to appease the tests in `core/ssm.test.ts`.
* We don't need this to contain anything useful because we're not testing the contents of it within that testing suite.
* */
