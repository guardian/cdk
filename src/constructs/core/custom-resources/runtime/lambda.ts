import { request } from "https";
import { parse } from "url";
// eslint-disable-next-line import/no-unresolved -- this comes from @types/aws-lambda, but eslint can't seem to read it properly
import type { CloudFormationCustomResourceEvent, Context } from "aws-lambda";
import SSM from "aws-sdk/clients/ssm";
import type { CustomResourceGetParameterProps } from "../interfaces";

/* eslint-disable -- This function is copied straight from AWS */
// https://github.com/aws/aws-cdk/blob/95438b56bfdc90e94f969f6998e5b5b680cbd7a8/packages/%40aws-cdk/custom-resources/lib/aws-custom-resource/runtime/index.ts#L16-L29
export function flatten(object: Record<string, unknown>): Record<string, string> {
  return Object.assign(
    {},
    ...(function _flatten(child: Record<string, any>, path: string[] = []): any {
      return [].concat(
        ...Object.keys(child).map((key) => {
          const childKey = Buffer.isBuffer(child[key]) ? child[key].toString() : child[key];
          return typeof childKey === "object"
            ? _flatten(childKey, path.concat([key]))
            : { [path.concat([key]).join(".")]: childKey };
        })
      );
    })(object)
  ) as Record<string, string>;
}
/* eslint-enable */

export async function handler(event: CloudFormationCustomResourceEvent, context: Context): Promise<void> {
  try {
    console.log(JSON.stringify(event));

    // Default physical resource id
    let physicalResourceId: string;
    switch (event.RequestType) {
      case "Create":
        physicalResourceId = event.LogicalResourceId;
        break;
      case "Update":
        physicalResourceId = event.PhysicalResourceId;
        break;
      case "Delete":
        await respond("SUCCESS", "OK", event.PhysicalResourceId, {});
        return;
    }

    let data: Record<string, string> = {};

    const getParamsProps = event.ResourceProperties.getParamsProps as string | undefined;

    if (getParamsProps) {
      const request = decodeCall(getParamsProps);
      const ssmClient = new SSM();
      const response = await ssmClient.getParameter(request.apiRequest).promise();
      console.log("Response:", JSON.stringify(response, null, 4));
      data = { ...flatten((response as unknown) as Record<string, string>) };
    }

    console.log("data: ", data);
    await respond("SUCCESS", "OK", physicalResourceId, data);
  } catch (e) {
    console.log(e);
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/message
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access -- e.message is valid, see docs above
    await respond("FAILED", e.message || "Internal Error", context.logStreamName, {});
  }

  function respond(responseStatus: string, reason: string, physicalResourceId: string, data: Record<string, string>) {
    const responseBody = JSON.stringify({
      Status: responseStatus,
      Reason: reason,
      PhysicalResourceId: physicalResourceId,
      StackId: event.StackId,
      RequestId: event.RequestId,
      LogicalResourceId: event.LogicalResourceId,
      NoEcho: false,
      Data: data,
    });

    console.log("Responding", responseBody);
    const parsedUrl = parse(event.ResponseURL);
    const requestOptions = {
      hostname: parsedUrl.hostname,
      path: parsedUrl.path,
      method: "PUT",
      headers: { "content-type": "", "content-length": responseBody.length },
    };

    return new Promise((resolve, reject) => {
      try {
        const r = request(requestOptions, resolve);
        r.on("error", reject);
        r.write(responseBody);
        r.end();
      } catch (e) {
        reject(e);
      }
    });
  }
}

function decodeCall(call: string): CustomResourceGetParameterProps {
  return JSON.parse(call) as CustomResourceGetParameterProps;
}
