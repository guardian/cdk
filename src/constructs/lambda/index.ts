export * from "./lambda";
export * from "./event-sources";
