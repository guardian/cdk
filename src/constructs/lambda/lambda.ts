import { Code, Function, RuntimeFamily } from "@aws-cdk/aws-lambda";
import type { FunctionProps, Runtime } from "@aws-cdk/aws-lambda";
import { Bucket } from "@aws-cdk/aws-s3";
import { Duration } from "@aws-cdk/core";
import { GuDistributable } from "../../types/distributable";
import { GuLambdaErrorPercentageAlarm } from "../cloudwatch";
import type { GuLambdaErrorPercentageMonitoringProps } from "../cloudwatch";
import { GuDistributionBucketParameter } from "../core";
import type { GuStack } from "../core";
import { AppIdentity } from "../core/identity";
import { GuParameterStoreReadPolicyStatement } from "../iam";

export interface GuFunctionProps extends GuDistributable, Omit<FunctionProps, "code">, AppIdentity {
  errorPercentageMonitoring?: GuLambdaErrorPercentageMonitoringProps;
}

function defaultMemorySize(runtime: Runtime, memorySize?: number): number {
  if (memorySize) {
    return memorySize;
  } else {
    switch (runtime.family) {
      case RuntimeFamily.JAVA:
        return 1024;
      default:
        return 512;
    }
  }
}

/**
 * Construct which creates a Lambda function.
 *
 * This lambda relies on the code artifact residing in a standard location in S3. For more details on the bucket used,
 * see [[`GuDistributionBucketParameter`]]. By default, the path used will be `stack/stage/app/fileName`.
 *
 * The default memory size of this Lambda will vary depending on the runtime chosen.
 * For Java runtimes (i.e. resource hungry Scala Lambdas!), 1024MB will be used. For all other runtimes,
 * the memory size defaults to 512MB. This can be overridden via the `memorySize` prop.
 *
 * By default, the timeout for this lambda is 30 seconds. This can be overridden via the `timeout` prop.
 *
 * By default the Lambda is granted permission to read from the SSM parameter store subtree specific to this lambda
 * (i.e. it can read all keys under <stage>/<stack>/<app>/). The lambda has STACK/STAGE/APP environment variables
 * which can be used to determine its identity.
 *
 * Note that this construct creates a Lambda without a trigger/event source. Depending on your use-case, you may wish to
 * consider using a pattern which instantiates a Lambda with a trigger e.g. [[`GuScheduledLambda`]].
 */
export class GuLambdaFunction extends Function {
  public readonly app: string;

  constructor(scope: GuStack, id: string, props: GuFunctionProps) {
    const { app, fileName, runtime, memorySize, timeout } = props;

    const defaultEnvironmentVariables = {
      STACK: scope.stack,
      STAGE: scope.stage,
      APP: app,
    };

    const bucket = Bucket.fromBucketName(
      scope,
      `${id}-bucket`,
      GuDistributionBucketParameter.getInstance(scope).valueAsString
    );
    const objectKey = GuDistributable.getObjectKey(scope, { app }, { fileName });
    const code = Code.fromBucket(bucket, objectKey);
    super(scope, id, {
      ...props,
      environment: {
        ...props.environment,
        ...defaultEnvironmentVariables,
      },
      memorySize: defaultMemorySize(runtime, memorySize),
      timeout: timeout ?? Duration.seconds(30),
      code,
    });

    this.app = app;

    if (props.errorPercentageMonitoring) {
      new GuLambdaErrorPercentageAlarm(scope, `${id}-ErrorPercentageAlarmForLambda`, {
        ...props.errorPercentageMonitoring,
        lambda: this,
      });
    }

    bucket.grantRead(this);

    const ssmParamReadPolicy = new GuParameterStoreReadPolicyStatement(scope, props);
    this.addToRolePolicy(ssmParamReadPolicy);

    AppIdentity.taggedConstruct(props, this);
  }
}
