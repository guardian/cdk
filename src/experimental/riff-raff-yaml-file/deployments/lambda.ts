import type { GuStack } from "../../../constructs/core";
import type { GuLambdaFunction } from "../../../constructs/lambda";
import type { RiffRaffDeployment } from "../types";

interface S3LocationProps {
  bucketSsmLookup?: boolean;
  bucketSsmKey?: string;
  prefixStackToKey?: boolean;
  prefixAppToKey?: boolean;
  prefixStageToKey?: boolean;
}

const locationProps = (lambda: GuLambdaFunction): S3LocationProps => {
  const bucketProp =
    lambda.bucketNamePath === undefined
      ? {
          bucketSsmLookup: true,
        }
      : {
          bucketSsmKey: lambda.bucketNamePath,
        };

  /**
   * If the lambda has a file prefix, we don't want to prefix the key with the stack, app or stage.
   * The default value for these properties is `true`, so we don't need to set them explicitly if
   * we *do* want to add the prefixes (and RiffRaff will complain if we add them).
   */
  const prefixProps = lambda.withoutFilePrefix
    ? {
        prefixStackToKey: false,
        prefixAppToKey: false,
        prefixStageToKey: false,
      }
    : {};

  return {
    ...bucketProp,
    ...prefixProps,
  };
};

export function uploadLambdaArtifact(lambda: GuLambdaFunction): RiffRaffDeployment {
  const { app, fileName } = lambda;
  const { stack, region } = lambda.stack as GuStack;

  return {
    name: ["lambda-upload", region, stack, app].join("-"),
    props: {
      type: "aws-lambda",
      stacks: new Set([stack]),
      regions: new Set([region]),
      app,
      contentDirectory: app,
      parameters: {
        ...locationProps(lambda),
        lookupByTags: true,
        fileName,
      },
      actions: ["uploadLambda"],
    },
  };
}

export function updateLambdaDeployment(
  lambda: GuLambdaFunction,
  { name: cfnDeployName }: RiffRaffDeployment,
): RiffRaffDeployment {
  const { app, fileName } = lambda;
  const { stack, region } = lambda.stack as GuStack;

  return {
    name: ["lambda-update", region, stack, app].join("-"),
    props: {
      type: "aws-lambda",
      stacks: new Set([stack]),
      regions: new Set([region]),
      app,
      contentDirectory: app,
      parameters: {
        ...locationProps(lambda),
        lookupByTags: true,
        fileName,
      },
      actions: ["updateLambda"],
      dependencies: [cfnDeployName],
    },
  };
}
