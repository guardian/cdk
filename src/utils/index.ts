export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
