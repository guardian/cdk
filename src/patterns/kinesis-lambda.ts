import type { StreamProps } from "@aws-cdk/aws-kinesis";
import { Stream, StreamEncryption } from "@aws-cdk/aws-kinesis";
import { StartingPosition } from "@aws-cdk/aws-lambda";
import type { KinesisEventSourceProps } from "@aws-cdk/aws-lambda-event-sources";
import { KinesisEventSource } from "@aws-cdk/aws-lambda-event-sources";
import type { GuLambdaErrorPercentageMonitoringProps, NoMonitoring } from "../constructs/cloudwatch";
import type { GuStack } from "../constructs/core";
import { AppIdentity } from "../constructs/core/identity";
import type { GuKinesisStreamProps } from "../constructs/kinesis";
import { GuKinesisStream } from "../constructs/kinesis";
import type { GuFunctionProps } from "../constructs/lambda";
import { GuLambdaFunction } from "../constructs/lambda";
import type { StreamErrorHandlingProps, StreamProcessingProps } from "../utils/lambda";
import { toAwsErrorHandlingProps } from "../utils/lambda";

/**
 * Used to provide information about an existing Kinesis stream to the [[`GuKinesisLambda`]] pattern.
 *
 * Specify a `logicalIdFromCloudFormation` to inherit a Kinesis stream which has already
 * been created via a CloudFormation stack. This is necessary to avoid data loss and interruptions of
 * service when migrating stacks from CloudFormation to `cdk`.
 *
 * Specify an `externalKinesisStreamName` to link the lambda to a Kinesis stream owned by a different stack
 * (or created outside of version control).
 *
 * **Example Usage**
 *
 * When migrating a CloudFormation stack which includes the following resource:
 * ```yaml
 * MyCloudFormedKinesisStream:
 *   Type: AWS::Kinesis::Stream
 * ```
 * Inherit the Kinesis stream (rather than creating a new one) using:
 * ```typescript
 * existingKinesisStream: { logicalIdFromCloudFormation: "MyCloudFormedKinesisStream" }
 * ```
 *
 * Alternatively, reference a Kinesis stream which belongs to another stack or pattern using:
 * ```typescript
 * existingKinesisStream: { externalKinesisStreamName: "KinesisStreamFromAnotherStack" }
 * ```
 */
export interface ExistingKinesisStream {
  logicalIdFromCloudFormation?: string;
  externalKinesisStreamName?: string;
}

/**
 * Configuration options for the [[`GuKinesisLambda`]] pattern.
 *
 * For all lambda function configuration options, see [[`GuFunctionProps`]].
 *
 * The `existingKinesisStream` property can be used to inherit or reference a Kinesis stream which
 * has been created outside of this pattern (i.e. via CloudFormation, or via a different `cdk` pattern, or stack).
 * For more details and example usage, see [[`ExistingKinesisStream`]].
 * If this property is omitted, the [[`GuKinesisLambda`]] pattern will create a new stream.
 *
 * If you have specific stream configuration requirements (e.g. data retention period), these can be set via
 * `kinesisStreamProps`.
 *
 * If you need to override the default stream processing options (e.g. batch size and parallelization), pass
 * [[`StreamProcessingProps`]] via `processingProps`.
 *
 * You must provide `errorHandlingConfiguration` to this pattern. Retry conditions can be configured
 * via [[`StreamErrorHandlingProps`]].
 *
 * It is advisable to configure an alarm based on the lambda's error percentage.
 * To do this, add the `monitoringConfiguration` property. The required properties for this are:
 *
 * ```typescript
 * monitoringConfiguration: {
 *   toleratedErrorPercentage: <sensible_error_percentage_threshold>,
 *   snsTopicName: "my-topic-for-cloudwatch-alerts",
 * }
 * ```
 * Other alarm properties (e.g. alarm name and description) will be pre-populated with sensible defaults.
 * For a full list of optional properties, see [[`GuLambdaErrorPercentageMonitoringProps`]].
 *
 * If your team do not use CloudWatch, it's possible to opt-out with the following configuration:
 * ```typescript
 *  monitoringConfiguration: { noMonitoring: true } as NoMonitoring
 * ```
 */
export interface GuKinesisLambdaProps extends Omit<GuFunctionProps, "rules" | "apis" | "errorPercentageMonitoring"> {
  monitoringConfiguration: NoMonitoring | GuLambdaErrorPercentageMonitoringProps;
  existingKinesisStream?: ExistingKinesisStream;
  errorHandlingConfiguration: StreamErrorHandlingProps;
  kinesisStreamProps?: StreamProps;
  processingProps?: StreamProcessingProps;
}

/**
 * Pattern which creates all of the resources needed to invoke a lambda function whenever a record is
 * put onto a Kinesis stream.
 *
 * This pattern will create a new Kinesis stream by default. If you are migrating a stack from CloudFormation,
 * you will need to opt-out of this behaviour. For information on overriding the default behaviour,
 * see [[`GuKinesisLambdaProps`]].
 */
export class GuKinesisLambda extends GuLambdaFunction {
  constructor(scope: GuStack, id: string, props: GuKinesisLambdaProps) {
    super(scope, id, {
      ...props,
      errorPercentageMonitoring: props.monitoringConfiguration.noMonitoring ? undefined : props.monitoringConfiguration,
    });
    const kinesisProps: GuKinesisStreamProps = {
      overrideId: !!props.existingKinesisStream?.logicalIdFromCloudFormation,
      encryption: StreamEncryption.MANAGED,
      ...props.kinesisStreamProps,
    };
    const streamId = props.existingKinesisStream?.logicalIdFromCloudFormation ?? "KinesisStream";

    const kinesisStream = props.existingKinesisStream?.externalKinesisStreamName
      ? Stream.fromStreamArn(
          scope,
          streamId,
          `arn:aws:kinesis:${scope.region}:${scope.account}:stream/${props.existingKinesisStream.externalKinesisStreamName}`
        )
      : AppIdentity.taggedConstruct(props, new GuKinesisStream(scope, streamId, kinesisProps));

    const errorHandlingPropsToAwsProps = toAwsErrorHandlingProps(props.errorHandlingConfiguration);

    const eventSourceProps: KinesisEventSourceProps = {
      startingPosition: StartingPosition.LATEST,
      ...props.processingProps,
      ...errorHandlingPropsToAwsProps,
    };
    this.addEventSource(new KinesisEventSource(kinesisStream, eventSourceProps));
  }
}
