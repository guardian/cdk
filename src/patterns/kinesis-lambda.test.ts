import { Duration } from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";
import type { StreamProps } from "aws-cdk-lib/aws-kinesis";
import { StreamEncryption } from "aws-cdk-lib/aws-kinesis";
import { Runtime, StartingPosition } from "aws-cdk-lib/aws-lambda";
import type { NoMonitoring } from "../constructs/cloudwatch";
import { StreamRetry } from "../utils/lambda";
import type { StreamErrorHandlingProps, StreamProcessingProps } from "../utils/lambda";
import { GuTemplate, simpleGuStackForTesting } from "../utils/test";
import { GuKinesisLambda } from "./kinesis-lambda";

describe("The GuKinesisLambda pattern", () => {
  it("should create the correct resources for a new stack with minimal config", () => {
    const stack = simpleGuStackForTesting();
    const basicErrorHandling: StreamErrorHandlingProps = {
      bisectBatchOnError: false,
      retryBehaviour: StreamRetry.maxAttempts(1),
    };
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      errorHandlingConfiguration: basicErrorHandling,
      monitoringConfiguration: noMonitoring,
      app: "testing",
    };
    new GuKinesisLambda(stack, "my-lambda-function", props);
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("should inherit an existing Kinesis stream correctly if an existingLogicalId is passed via existingSnsTopic in a migrating stack", () => {
    const stack = simpleGuStackForTesting();
    const basicErrorHandling: StreamErrorHandlingProps = {
      bisectBatchOnError: false,
      retryBehaviour: StreamRetry.maxAttempts(1),
    };
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      errorHandlingConfiguration: basicErrorHandling,
      monitoringConfiguration: noMonitoring,
      app: "testing",
    };
    const { kinesisStream } = new GuKinesisLambda(stack, "my-lambda-function", props);
    stack.overrideLogicalId(kinesisStream, { logicalId: "pre-existing-kinesis-stream", reason: "testing" });
    GuTemplate.fromStack(stack).hasResourceWithLogicalId("AWS::Kinesis::Stream", "pre-existing-kinesis-stream");
  });

  it("should not generate a new Kinesis stream if an external stream name is passed in", () => {
    const stack = simpleGuStackForTesting();
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const basicErrorHandling: StreamErrorHandlingProps = {
      bisectBatchOnError: false,
      retryBehaviour: StreamRetry.maxAttempts(1),
    };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      monitoringConfiguration: noMonitoring,
      errorHandlingConfiguration: basicErrorHandling,
      existingKinesisStream: { externalKinesisStreamName: "kinesis-stream-from-another-stack" },
      app: "testing",
    };
    new GuKinesisLambda(stack, "my-lambda-function", props);

    const template = Template.fromStack(stack);
    template.hasResourceProperties("AWS::Lambda::EventSourceMapping", {
      EventSourceArn: {
        "Fn::Join": [
          "",
          [
            "arn:aws:kinesis:",
            {
              Ref: "AWS::Region",
            },
            ":",
            {
              Ref: "AWS::AccountId",
            },
            ":stream/kinesis-stream-from-another-stack",
          ],
        ],
      },
    });
    template.resourceCountIs("AWS::Kinesis::Stream", 0);
  });

  it("should create an alarm if monitoring configuration is provided", () => {
    const stack = simpleGuStackForTesting();
    const basicErrorHandling: StreamErrorHandlingProps = {
      bisectBatchOnError: false,
      retryBehaviour: StreamRetry.maxAttempts(1),
    };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      errorHandlingConfiguration: basicErrorHandling,
      monitoringConfiguration: {
        toleratedErrorPercentage: 1,
        snsTopicName: "alerts-topic",
      },
      app: "testing",
    };
    new GuKinesisLambda(stack, "my-lambda-function", props);
    Template.fromStack(stack).resourceCountIs("AWS::CloudWatch::Alarm", 1);
  });

  it("should use managed encryption by default", () => {
    const stack = simpleGuStackForTesting();
    const basicErrorHandling: StreamErrorHandlingProps = {
      bisectBatchOnError: false,
      retryBehaviour: StreamRetry.maxAttempts(1),
    };
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      errorHandlingConfiguration: basicErrorHandling,
      monitoringConfiguration: noMonitoring,
      app: "testing",
    };
    new GuKinesisLambda(stack, "my-lambda-function", props);
    Template.fromStack(stack).hasResourceProperties("AWS::Kinesis::Stream", {
      StreamEncryption: {
        EncryptionType: "KMS",
        KeyId: "alias/aws/kinesis",
      },
    });
  });

  it("should override the default encryption type if encryption is explicitly set", () => {
    const stack = simpleGuStackForTesting();
    const basicErrorHandling: StreamErrorHandlingProps = {
      bisectBatchOnError: false,
      retryBehaviour: StreamRetry.maxAttempts(1),
    };
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      errorHandlingConfiguration: basicErrorHandling,
      monitoringConfiguration: noMonitoring,
      kinesisStreamProps: {
        encryption: StreamEncryption.UNENCRYPTED,
      },
      app: "testing",
    };
    new GuKinesisLambda(stack, "my-lambda-function", props);

    Template.fromStack(stack).hasResourceProperties("AWS::Kinesis::Stream", {
      StreamEncryption: Match.absent(),
    });
  });

  it("should pass through other Kinesis stream properties", () => {
    const stack = simpleGuStackForTesting();
    const basicErrorHandling: StreamErrorHandlingProps = {
      bisectBatchOnError: false,
      retryBehaviour: StreamRetry.maxAttempts(1),
    };
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const kinesisStreamProps: StreamProps = {
      streamName: "custom-kinesis-stream-name",
      retentionPeriod: Duration.hours(100),
      shardCount: 3,
    };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      errorHandlingConfiguration: basicErrorHandling,
      monitoringConfiguration: noMonitoring,
      kinesisStreamProps: kinesisStreamProps,
      app: "testing",
    };
    new GuKinesisLambda(stack, "my-lambda-function", props);
    Template.fromStack(stack).hasResourceProperties("AWS::Kinesis::Stream", {
      Name: "custom-kinesis-stream-name",
      RetentionPeriodHours: 100,
      ShardCount: 3,
    });
  });

  it("should pass through processing properties", () => {
    const stack = simpleGuStackForTesting();
    const basicErrorHandling: StreamErrorHandlingProps = {
      bisectBatchOnError: false,
      retryBehaviour: StreamRetry.maxAttempts(1),
    };
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const processingProps: StreamProcessingProps = {
      batchSize: 11,
      parallelizationFactor: 2,
      startingPosition: StartingPosition.TRIM_HORIZON,
      maxBatchingWindow: Duration.minutes(3),
      enabled: false,
    };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      errorHandlingConfiguration: basicErrorHandling,
      monitoringConfiguration: noMonitoring,
      processingProps: processingProps,
      app: "testing",
    };
    new GuKinesisLambda(stack, "my-lambda-function", props);
    Template.fromStack(stack).hasResourceProperties("AWS::Lambda::EventSourceMapping", {
      BatchSize: 11,
      ParallelizationFactor: 2,
      StartingPosition: "TRIM_HORIZON",
      MaximumBatchingWindowInSeconds: 180,
      Enabled: false,
    });
  });

  it("should configure error handling correctly based on max number of retry attempts", () => {
    const stack = simpleGuStackForTesting();
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const errorHandlingProps: StreamErrorHandlingProps = {
      bisectBatchOnError: true,
      retryBehaviour: StreamRetry.maxAttempts(5),
    };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      errorHandlingConfiguration: errorHandlingProps,
      monitoringConfiguration: noMonitoring,
      app: "testing",
    };
    new GuKinesisLambda(stack, "my-lambda-function", props);
    Template.fromStack(stack).hasResourceProperties("AWS::Lambda::EventSourceMapping", {
      BisectBatchOnFunctionError: true,
      MaximumRetryAttempts: 5,
    });
  });

  it("should configure error handling correctly based on max age of records", () => {
    const stack = simpleGuStackForTesting();
    const noMonitoring: NoMonitoring = { noMonitoring: true };
    const errorHandlingProps: StreamErrorHandlingProps = {
      bisectBatchOnError: true,
      retryBehaviour: StreamRetry.maxAge(Duration.minutes(5)),
    };
    const props = {
      fileName: "lambda.zip",
      functionName: "my-lambda-function",
      handler: "my-lambda/handler",
      runtime: Runtime.NODEJS_12_X,
      errorHandlingConfiguration: errorHandlingProps,
      monitoringConfiguration: noMonitoring,
      app: "testing",
    };
    new GuKinesisLambda(stack, "my-lambda-function", props);
    Template.fromStack(stack).hasResourceProperties("AWS::Lambda::EventSourceMapping", {
      BisectBatchOnFunctionError: true,
      MaximumRecordAgeInSeconds: 300,
    });
  });
});
