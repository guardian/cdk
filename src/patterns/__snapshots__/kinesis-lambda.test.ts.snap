// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuKinesisLambda pattern should create the correct resources for a new stack with minimal config 1`] = `
Object {
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "KinesisStream46752A3E": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "mylambdafunction8D341B54": Object {
      "DependsOn": Array [
        "mylambdafunctionServiceRoleDefaultPolicy769897D4",
        "mylambdafunctionServiceRoleE82C2E25",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "lambda.zip",
        },
        "FunctionName": "my-lambda-function",
        "Handler": "my-lambda/handler",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "mylambdafunctionServiceRoleE82C2E25",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mylambdafunctionKinesisEventSourceTestKinesisStream4CAC6550FA5E175D": Object {
      "Properties": Object {
        "BatchSize": 100,
        "BisectBatchOnFunctionError": false,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "KinesisStream46752A3E",
            "Arn",
          ],
        },
        "FunctionName": Object {
          "Ref": "mylambdafunction8D341B54",
        },
        "MaximumRetryAttempts": 1,
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "mylambdafunctionServiceRoleDefaultPolicy769897D4": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "KinesisStream46752A3E",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "kinesis:DescribeStream",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "KinesisStream46752A3E",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mylambdafunctionServiceRoleDefaultPolicy769897D4",
        "Roles": Array [
          Object {
            "Ref": "mylambdafunctionServiceRoleE82C2E25",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "mylambdafunctionServiceRoleE82C2E25": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "testing",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
